### PR TITLE
Improve error message for changing tensor metadata after .data or .detach()

### PR DESCRIPTION
**Pull request by yf225**
When a user tries to change tensor metadata after `.data` or `.detach()`, we currently shows an error message "<function_name> is not allowed on Tensor created from .data or .detach()". However, this error message is not informative enough in that it doesn't suggest what the right fix should be (which leads to user posts such as https://discuss.pytorch.org/t/api-change-for-tensor-data-set-in-torch-nightly/33310). This PR improves the error message.

### CODE REVIEW

File: aten/src/ATen/SparseTensorImpl.cpp [hunk 1]
```diff
@@ -79,7 +79,9 @@ int64_t SparseTensorImpl::storage_offset() const {
   AT_ERROR("sparse tensors do not have storage");
 }
 void SparseTensorImpl::set_indices_and_values_unsafe(const Tensor& indices, const Tensor& values) {
-  AT_CHECK(allow_tensor_metadata_change(), "set_indices_and_values_unsafe is not allowed on Tensor created from .data or .detach()");
+  AT_CHECK(allow_tensor_metadata_change(), "set_indices_and_values_unsafe is not allowed on Tensor created from .data or .detach(). "
+    "If you are using Python API, please make sure the Python function is called on the original tensor (not the tensor created from .data or .detach()), "
+    "and the function call is wrapped in a `with torch.no_grad():` block");
```
Comment (ezyang) [inline aten/src/ATen/SparseTensorImpl.cpp]: I think it would be better if we could declare the instructions once, instead of copy pasting it this many places, in case we want to edit in the future

File: aten/src/ATen/SparseTensorImpl.cpp [hunk 2]
```diff
@@ -79,7 +79,7 @@ int64_t SparseTensorImpl::storage_offset() const {
   AT_ERROR("sparse tensors do not have storage");
 }
 void SparseTensorImpl::set_indices_and_values_unsafe(const Tensor& indices, const Tensor& values) {
-  AT_CHECK(allow_tensor_metadata_change(), "set_indices_and_values_unsafe is not allowed on Tensor created from .data or .detach()");
+  AT_CHECK(allow_tensor_metadata_change(), "set_indices_and_values_unsafe " + FLAGS_err_msg_for_changing_metadata_on_derived_tensor);
```
Comment (ezyang) [inline aten/src/ATen/SparseTensorImpl.cpp]: FYI you don't need to do a `std::string` append; `AT_CHECK` is variadic and will concat each of the arguments together.

File: c10/core/TensorImpl.cpp [hunk 3]
```diff
@@ -15,6 +15,14 @@ C10_DEFINE_int64(
     "The maximum memory in bytes to keep on shrink, if the difference between "
     "tensor sizes is bigger than this then tensor will be reset.");
 
+C10_DEFINE_string(
```
Comment (ezyang) [inline c10/core/TensorImpl.cpp]: Uhhhhhh... I would *not* make this a gflag lol. (That's what `C10_DEFINE_string` does; it doesn't create a string constant, if that's what you thought it does...)

**Review by ezyang — CHANGES_REQUESTED** [review]
That is a... very strange way of defining a string constant lol

File: c10/core/TensorImpl.cpp [hunk 4]
```diff
@@ -15,6 +15,12 @@ C10_DEFINE_int64(
     "The maximum memory in bytes to keep on shrink, if the difference between "
     "tensor sizes is bigger than this then tensor will be reset.");
 
+const std::string err_msg_for_changing_metadata_on_derived_tensor =
+    "is not allowed on Tensor created from .data or .detach(). "
+    "If you are using Python API, please make sure the Python function is called on the original tensor "
```
Comment (gchanan) [inline c10/core/TensorImpl.cpp]: can this happen outside of the Python API?

File: c10/core/TensorImpl.cpp [hunk 5]
```diff
@@ -15,6 +15,12 @@ C10_DEFINE_int64(
...
+    "(not the tensor created from .data or .detach()), "
+    "and the function call is wrapped in a `with torch.no_grad():` block";
```
Comment (gchanan) [inline c10/core/TensorImpl.cpp]: I think you should just give a simple example, e.g. show set_.

File: aten/src/ATen/SparseTensorImpl.cpp [hunk 6]
```diff
@@ -79,7 +79,7 @@ int64_t SparseTensorImpl::storage_offset() const {
   AT_ERROR("sparse tensors do not have storage");
 }
 void SparseTensorImpl::set_indices_and_values_unsafe(const Tensor& indices, const Tensor& values) {
-  AT_CHECK(allow_tensor_metadata_change(), "set_indices_and_values_unsafe is not allowed on Tensor created from .data or .detach()");
+  AT_CHECK(allow_tensor_metadata_change(), "set_indices_and_values_unsafe ", err_msg_for_changing_metadata_on_derived_tensor);
```
Comment (gchanan) [inline aten/src/ATen/SparseTensorImpl.cpp]: I don't think you should start with the function name; the issue here is that you start the message with something that the user knows nothing about (what's set_indices_and_values_unsafe, that's not exposed anywhere in the normal Python or C++ API).

**Review by gchanan — CHANGES_REQUESTED** [review]
